aws_topology/stackstate_checks/aws_topology/cloudtrail.py

```diff
@@ -0,0 +1,126 @@
+                    if dt < not_before:
+                        to_delete.append({"Key": itm["Key"]})
+                        if len(to_delete) > 999:
+                            self._delete_files(client, bucket_name, to_delete)
+                            to_delete = []
+                    else:
+                        files_to_handle.append(itm["Key"])
+
+        self._delete_files(client, bucket_name, to_delete)
+        return self._process_files(client, bucket_name, files_to_handle)
+
+    def _get_messages_from_cloudtrail(self, not_before):
+        client = self.session.client("cloudtrail")
+        # collect the events (ordering is most recent event first)
+        for pg in client.get_paginator("lookup_events").paginate(
+            LookupAttributes=[{"AttributeKey": "ReadOnly", "AttributeValue": "false"}],
+        ):
+            for itm in pg.get("Events") or []:
+                rec = json.loads(itm["CloudTrailEvent"])
+                event_date = dateutil.parser.isoparse(rec["eventTime"])
+                if event_date > not_before:
+                    yield rec
+
+    def _delete_files(self, client, bucket_name, files):
+        try:
+            for i in range(0, len(files), 999):
+                client.delete_objects(
+                    Bucket=bucket_name,
+                    Delete={"Objects": files[i : i + 999], "Quiet": True},
+                )
+        except Exception:
+            self.agent.warning("CloudtrailCollector: Deleting s3 files failed")
+
+    def _is_gz_file(self, body):
+        with io.BytesIO(body) as test_f:
+            return test_f.read(2) == b"\x1f\x8b"
+
+    def _get_stream(self, body):
+        if isinstance(body, string_types):
+            # this case is only for test purposes
+            if PY2:
+                body = bytes(body)
+            else:
+                body = bytes(body, "ascii")
+        elif isinstance(body, botocore.response.StreamingBody):
+            body = body.read()
+        if self._is_gz_file(body):
+            return gzip.GzipFile(fileobj=io.BytesIO(body), mode="rb")
+        else:
+            return io.BytesIO(body)
+
+    def _process_files(self, client, bucket_name, files):
+        for file in reversed(files):
+            objects = []
+            s3_body = client.get_object(Bucket=bucket_name, Key=file).get("Body")
+            with self._get_stream(s3_body) as data:
+                decoder = json.JSONDecoder()
+                txt = data.read().decode("ascii")
+                while txt:
+                    try:
+                        obj, index = decoder.raw_decode(txt)
+                        txt = txt[index:]
+                        if "detail" in obj:
+                            objects.append(obj["detail"])
+                    except JSONParseException:
+                        txt = ""
+                for event in reversed(objects):
+                    yield event
```

Comment:
This wouldn't make too much of a difference, but you can either yield the events after the stream is closed or move the objects list declaration into the `with`. Now it's sitting outside which makes this look a bit strange